endless/eosprofile.c

```diff
@@ -0,0 +1,567 @@
+}
+
+static void
+eos_profile_probe_free (EosProfileProbe *probe)
+{
+  G_LOCK (profile_state);
+  g_hash_table_remove (profile_state->probes, probe->name);
+  G_UNLOCK (profile_state);
+
+  g_array_unref (probe->samples);
+  g_free (probe->name);
+  g_free (probe->function);
+  g_free (probe->file);
+
+  g_free (probe);
+}
+
+static EosProfileProbe *
+eos_profile_probe_ref (EosProfileProbe *probe)
+{
+  if (probe == &eos_profile_dummy_probe)
+    return probe;
+
+  g_atomic_int_inc (&probe->ref_count);
+}
+
+static void
+eos_profile_probe_unref (EosProfileProbe *probe)
+{
+  if (probe == &eos_profile_dummy_probe)
+    return;
+
+  if (g_atomic_int_dec_and_test (&probe->ref_count))
+    eos_profile_probe_stop (probe);
+}
+
+G_DEFINE_BOXED_TYPE (EosProfileProbe, eos_profile_probe,
+                     eos_profile_probe_ref,
+                     eos_profile_probe_unref)
+
+/**
+ * eos_profile_probe_start:
+ * @file: the source file for the probe, typically represented by %__FILE__
+ * @line: the line in the source @file, typically represented by %__LINE__
+ * @function: the function for the probe, typically represented by %G_STRFUNC
+ * @name: a unique name for the probe
+ *
+ * Starts a profiling probe for @name, creating it if necessary.
+ *
+ * Returns: (transfer none): a profile probe identifier; use eos_profile_probe_stop()
+ *   to stop the profiling on the returned probe
+ *
+ * Since: 0.6
+ */
+EosProfileProbe *
+eos_profile_probe_start (const char *file,
+                         gsize       line,
+                         const char *function,
+                         const char *name)
+{
+  /* Don't measure the lock */
+  gint64 sample_time = g_get_monotonic_time ();
+
+  /* We can take this out of the lock because by the time we reach
+   * eos_profile_probe_stop() the profile state is guaranteed to
+   * either always exist, or not
+   */
+  if (profile_state == NULL)
+    return &eos_profile_dummy_probe;
+
+  G_LOCK (profile_state);
+
+  EosProfileProbe *res = g_hash_table_lookup (profile_state->probes, name);
+  if (res == NULL)
+    {
+      res = eos_profile_probe_new (file, line, function, name);
+
+      g_hash_table_insert (profile_state->probes, res->name, res);
+    }
+
+  res->is_running = TRUE;
+
+  g_array_append_vals (res->samples,
+                       &(ProfileSample) {
+                         .start_time = sample_time,
+                         .end_time = -1
+                       },
+                       1);
+
+  G_UNLOCK (profile_state);
+
+  return (EosProfileProbe *) res;
+}
+
+/**
+ * eos_profile_probe_stop:
+ * @probe: a #EosProfileProbe
+ *
+ * Stops a profiling probe started using eos_profile_probe_start().
+ *
+ * Since: 0.6
+ */
+void
+eos_profile_probe_stop (EosProfileProbe *probe)
+{
+  if (probe == &eos_profile_dummy_probe)
+    return;
+
+  /* Don't measure the lock */
+  gint64 sample_time = g_get_monotonic_time ();
+
+  g_autoptr(GMutexLocker) locker = g_mutex_locker_new (&probe->probe_lock);
```

Comment:
Probes in JS are still not working for me. I get a segfault when locking the mutex at this line:
```
#0  0x00007f2d15b12425 in g_mutex_lock (mutex=mutex@entry=0x7f2ce36a8150 <eos_profile_probe_get_type+16>) at gthread-posix.c:1336
#1  0x00007f2ce36a82fb in eos_profile_probe_stop (mutex=0x7f2ce36a8150 <eos_profile_probe_get_type+16>) at /usr/include/glib-2.0/glib/gthread.h:321
#2  0x00007f2ce36a82fb in eos_profile_probe_stop (probe=0x7f2ce36a8120 <eos_profile_probe_ref>) at endless/eosprofile.c:262
#3  0x0000003997e05ed8 in ffi_call_unix64 ()
    at ../libffi-3.2.1/src/x86/unix64.S:76
#4  0x0000003997e0593a in ffi_call (cif=cif@entry=0x34b5658, fn=<optimized out>, rvalue=<optimized out>, 
    rvalue@entry=0x7ffcb11cd330, avalue=avalue@entry=0x7ffcb11cd210)
    at /usr/src/debug/libffi/3.2.1-r0/libffi-3.2.1/src/x86/ffi64.c:525
#5  0x00007f2d157dbd69 in gjs_invoke_c_function(JSContext*, Function*, JS::HandleObject, JS::HandleValueArray const&, mozilla::Maybe<JS::MutableHandle<JS::Value> >, GIArgument*) (context=context@entry=0x245b000, function=function@entry=0x34b5640, obj=obj@entry=..., args=..., js_rval=..., r_value=r_value@entry=0x0)
    at gi/function.cpp:1033
#6  0x00007f2d157dd582 in function_call(JSContext*, unsigned int, JS::Value*) (context=0x245b000, js_argc=0, vp=0x2695e30) at gi/function.cpp:1353
#7  0x00007f2d123fec50 in  () at /lib/libmozjs-52.so
#8  0x00007f2d123f96b1 in  () at /lib/libmozjs-52.so
#9  0x00007f2d123fe834 in  () at /lib/libmozjs-52.so
#10 0x00007f2d123fea82 in  () at /lib/libmozjs-52.so
#11 0x00007f2d123fed68 in  () at /lib/libmozjs-52.so
#12 0x00007f2d12041507 in  () at /lib/libmozjs-52.so
#13 0x00007f2d123feb26 in  () at /lib/libmozjs-52.so
#14 0x00007f2d123fed68 in  () at /lib/libmozjs-52.so
#15 0x00007f2d1228a48e in JS::Call(JSContext*, JS::Handle<JS::Value>, JS::Handle<JS::Value>, JS::HandleValueArray const&, JS::MutableHandle<JS::Value>) ()
    at /lib/libmozjs-52.so
#16 0x00007f2d157f797b in _gjs_context_run_jobs(GjsContext*) (rval=..., args=..., funObj=..., thisv=..., cx=0x245b000) at /usr/include/mozjs-52/jsapi.h:3330
#17 0x00007f2d157f797b in _gjs_context_run_jobs(GjsContext*) (gjs_context=gjs_context@entry=0x245a000 [GjsContext]) at gjs/context.cpp:568
#18 0x00007f2d157f7ac9 in drain_job_queue_idle_handler(void*) (data=0x245a000)
    at gjs/context.cpp:490
#19 0x00007f2d15acc2da in g_main_context_dispatch (context=0x26ef920)
    at gmain.c:3148
#20 0x00007f2d15acc2da in g_main_context_dispatch (context=context@entry=0x26ef920) at gmain.c:3813
#21 0x00007f2d15acc698 in g_main_context_iterate (context=context@entry=0x26ef920, block=block@entry=1, dispatch=dispatch@entry=1, self=<optimized out>)
    at gmain.c:3886
#22 0x00007f2d15acc74c in g_main_context_iteration (context=context@entry=0x26ef920, may_block=may_block@entry=1) at gmain.c:3947
#23 0x00007f2d1608797d in g_application_run (application=0x27f6130 [EknApplication], argc=2, argv=0x27422a0) at gapplication.c:2401
#24 0x0000003997e05ed8 in ffi_call_unix64 ()
    at ../libffi-3.2.1/src/x86/unix64.S:76
#25 0x0000003997e0593a in ffi_call (cif=cif@entry=0x273d998, fn=<optimized out>, rvalue=<optimized out>, 
    rvalue@entry=0x7ffcb11ce660, avalue=avalue@entry=0x7ffcb11ce520)
    at /usr/src/debug/libffi/3.2.1-r0/libffi-3.2.1/src/x86/ffi64.c:525
#26 0x00007f2d157dbd69 in gjs_invoke_c_function(JSContext*, Function*, JS::HandleObject, JS::HandleValueArray const&, mozilla::Maybe<JS::MutableHandle<JS::Value> >, GIArgument*) (context=context@entry=0x245b000, function=function@entry=0x273d980, obj=..., 
    obj@entry=..., args=..., js_rval=..., r_value=r_value@entry=0x0)
    at gi/function.cpp:1033
#27 0x00007f2d157dd582 in function_call(JSContext*, unsigned int, JS::Value*) (context=0x245b000, js_argc=1, vp=0x2695da0) at gi/function.cpp:1353
#28 0x00007f2d123fec50 in  () at /lib/libmozjs-52.so
#29 0x00007f2d123f96b1 in  () at /lib/libmozjs-52.so
#30 0x00007f2d123fe834 in  () at /lib/libmozjs-52.so
#31 0x00007f2d12404a9a in  () at /lib/libmozjs-52.so
#32 0x00007f2d12404bb7 in  () at /lib/libmozjs-52.so
#33 0x00007f2d12277aeb in  () at /lib/libmozjs-52.so
#34 0x00007f2d15805698 in gjs_eval_with_scope(JSContext*, JS::HandleObject, char const*, ssize_t, char const*, JS::MutableHandleValue) (context=0x245b000, object=object@entry=..., script=<optimized out>, 
    script@entry=0x2455a00 "const Gio = imports.gi.Gio;\nGio.Resource.load('/usr/share/eos-knowledge/eos-knowledge.gresource')._register();\nimports.searchPath.unshift('resource:///com/endlessm/knowledge/js');\n\nconst Application ="..., script_len=script_len@entry=460, filename=filename@entry=0x401ca4 "<command line>", retval=retval@entry=...) at gjs/jsapi-util.cpp:876
#35 0x00007f2d157f7c07 in gjs_context_eval(GjsContext*, char const*, gssize, char const*, int*, GError**) (js_context=js_context@entry=0x245a000 [GjsContext], script=0x2455a00 "const Gio = imports.gi.Gio;\nGio.Resource.load('/usr/share/eos-knowledge/eos-knowledge.gresource')._register();\nimports.searchPath.unshift('resource:///com/endlessm/knowledge/js');\n\nconst Application ="..., script_len=46---Type <return> to continue, or q <return> to quit---
0, filename=filename@entry=0x401ca4 "<command line>", exit_status_p=exit_status_p@entry=0x7ffcb11cf520, error=error@entry=0x7ffcb11cf4d8)
    at gjs/context.cpp:701
#36 0x00000000004019a1 in main(int, char**) (argc=<optimized out>, argv=<optimized out>) at gjs/console.cpp:284
```

Reply:
In light of that, it might be good to add a regression test to eos-sdk's JS tests, for using the probe from JS code.

Reply:
The issue was the `transfer none` annotation in the introspection, which led to a double unref/double stop.

The memory management is a bit wonky, though, because we're not returning a real object, but a boxed type.

The real fix is probably making EosProfileProbe a GObject.

Reply:
> The real fix is probably making EosProfileProbe a GObject.

Nope, the real fix was not to implement memory management semantics on top of something that is already managed. :-)